### PR TITLE
PD-553: Write basic dRAID primer

### PR DESCRIPTION
**Pull request by Mrt134**
This unversioned background information article builds on some internal investigation into dRAID and adapts portions of the OpenZFS documentation on the feature.

Thanks for contributing to TrueNAS documentation! By opening a Pull Request, you're acknowledging that your changes will be distributed under the [Creative Commons 4.0](https://creativecommons.org/licenses/by-nc-sa/4.0/) license.

### CODE REVIEW

File: content/References/dRAIDprimer.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+It is recommended to review this list of terms and definitions before attempting to create storage pools with a dRAID layout in TrueNAS:
+
+{{< truetable >}}
+| Term | Definition |
+|------|------------|
+| Children (**C**) | Number of drives included in the dRAID deployment.
+| Data Devices (**D**) | Number of data devices in a redundancy group. This number can be quite high, but generally a lower number results in greater performances and capacity is more effectively used. |
+| Distributed hot spare | Unlike in a raidz configuration where spares remain inactive until needed, in a dRAID configuration spare capacity is distributed across the drives. This results in all drives being active members of the pool.
+| Parity Level (**P**) | Distributed parity level of a dRAID redundancy group. This ranges from **1** to **3** and is similar to the RAIDz parity level. |
+| Redundancy group | Similar to a **vdev** in raidz. A redundancy group is composed of parity devices and data devices. Redundancy group size impacts storage performance and capacity. |
```
Comment (DjP-iX) [inline content/References/dRAIDprimer.md:87]: This definition should be scrubbed to diferentiate it from Vdev below. Maybe just cut the first sentence--or explain in what way it's similar to vdevs in raidz

File: content/References/dRAIDprimer.md [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+The terminology used for a dRAID layout is similar to that used for raidz, but could have a different meaning.
+It is recommended to review this list of terms and definitions before attempting to create storage pools with a dRAID layout in TrueNAS:
+
+{{< truetable >}}
+| Term | Definition |
+|------|------------|
+| Children (**C**) | Number of drives included in the dRAID deployment.
+| Data Devices (**D**) | Number of data devices in a redundancy group. This number can be quite high, but generally a lower number results in greater performances and capacity is more effectively used. |
+| Distributed hot spare | Unlike in a raidz configuration where spares remain inactive until needed, in a dRAID configuration spare capacity is distributed across the drives. This results in all drives being active members of the pool.
+| Parity Level (**P**) | Distributed parity level of a dRAID redundancy group. This ranges from **1** to **3** and is similar to the RAIDz parity level. |
```
Comment (DjP-iX) [inline content/References/dRAIDprimer.md:86]:
```suggestion
| Parity Level (**P**) | Distributed parity level of a dRAID redundancy group. This ranges from **1** to **3** and is similar to the raidz parity level. |
```

File: content/References/dRAIDprimer.md [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+## Concepts
+
+**Healing resilver**
+
+The traditional ZFS resilver.
+The entire block tree is scanned and traversed.
+Checksums are verified during the repair process.
+This can be a slow process as it results in a largely random workload that is not ideal for performance.
+In a RAIDz deployment this also puts extra strain on the remaining discs in the vdev, as they are all being read from simultaneously.
```
Comment (DjP-iX) [inline content/References/dRAIDprimer.md:33]:
```suggestion
In a raidz deployment this also puts extra strain on the remaining discs in the vdev, as they are all being read from simultaneously.
```

File: content/References/dRAIDprimer.md [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+description: "Background information and general information about the dRAID storage solution included in OpenZFS version 2.1.0."
+weight: 72
+---
+
+{{< toc >}}
+
+Introduced in OpenZFS version 2.1.0 and initially supported in TrueNAS SCALE 23.10 (Cobia), declustered RAID (dRAID) is an alternate method for creating OpenZFS data virtual devices (vdevs).
+
+Intended for storage arrays with numerous attached disks (>100), the primary benefit of a dRAID vdev layout is to reduce resilver times.
+It does this by building the dRAID vdev from multiple internal raidz groups that have their own data and parity devices, using precomputed permutation maps for the rebuild IO, and using a fixed stripe width that fills storage capacity with zeroes when necessary.
```
Comment (william-gr) [inline content/References/dRAIDprimer.md:12]: `internal raidz groups`, i think you meant raid, not raidz

File: content/References/dRAIDprimer.md [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+The terminology used for a dRAID layout is similar to that used for raidz, but could have a different meaning.
+It is recommended to review this list of terms and definitions before attempting to create storage pools with a dRAID layout in TrueNAS:
+
+{{< truetable >}}
+| Term | Definition |
+|------|------------|
+| Children (**C**) | Number of drives included in the dRAID deployment. |
+| Data Devices (**D**) | Number of data devices in a redundancy group. This number can be quite high, but generally a lower number results in greater performances and capacity is more effectively used. |
+| Distributed hot spare (**S**) | Unlike in a raidz configuration where spares remain inactive until needed, in a dRAID configuration spare capacity is distributed across the drives. This results in all drives being active members of the pool. This number cannot change after pool creation. |
```
Comment (william-gr) [inline content/References/dRAIDprimer.md:91]: I wonder if a section about Spares would be worthwhile?
Spares work very differently in draid and one absolutely do not want to use traditional spares and these cannot be added after the vdev is created.

File: content/References/dRAIDprimer.md [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+Stripe width is fixed in dRAID, with zeroes added as padding.
+This can have a significant effect on usable capacity, depending on the data stored.
+
+In a redundancy group of eight data disks using 4k sector disks, the minimum allocation size is **32k**.
+Any files smaller than 32k still fill an entire stripe, with zeroes appended to the write to fill the entire stripe.
+This greatly reduces the pool usable capacity when the pool stores large numbers of small files.
+
+dRAID datasets benefit greatly from larger **record sizes**.
+**128k** is recommended as a minimum value for datasets and **64k** to **128k** for zvols.
```
Comment (william-gr) [inline content/References/dRAIDprimer.md:70]: That will be the very minimum. Wonder if we should mention reasonable defaults, like 1M for datasets to achieve better compression. Might want to consult with mav.

Comment (Mrt134) [reply content/References/dRAIDprimer.md]: I spoke with Mav and updated this content with some more detail based on expected I/O use for datasets/zvols. Let me know if that is ok or if we're getting too "in the weeds" here. Thanks!

**Review by micjohnson777 — COMMENTED** [review]
I'm questioning the use of bold for emphasis. This is outside of our style guide. We use italics for the first instance of a new term we define, bold is limited to UI ref elements.
Do these instances need this much emphasis? Also, use of bold in the tables when mentioning the values or letters for terms. If we do want to use bold, we need to update the style guide.

**Review by DjP-iX — APPROVED** [review]

**Pull request merged**